### ISSUE TITLE
Selenium graphql check doesn't respond back after few hours of running tests

### ISSUE THREAD

**Issue by yasha-dev1** [SeleniumHQ/docker-selenium#1318]
## 🐛 Bug Report

we are using Selenium grid4 for scheduling a lot of processes. it works perfectly but after a few days of running tests, selenium Grid4 dies. Simply the POST request `selenium-domain:4444/graphql` doesn't respond back at all. (see the below screenshot). I am using the below config. as seen I have added a liveness and readiness probe for K8s to handle if a node has a problem in processing. 

## To Reproduce

running tests after a quite long amount of time, selenium grid UI stops showing information.
the problem can get resolved by restarting all the selenium nodes(not only router). But this should not happen.

## Expected behavior

the POST request `selenium-domain:4444/graphql` respond in time.

![image](https://user-images.githubusercontent.com/58387199/122528610-ce13d500-d01c-11eb-84d9-ca340e418af1.png)

## K8s configuration I'm using
```` 
```
# To execute this yml file use this command in your local kubernetes environment `kubectl apply -f k8s-deployment-full-grid.yaml`
#***************************************
#SELENIUM EVENT BUS
#***************************************
apiVersion: v1
kind: Service
metadata:
  name: selenium-event-bus
  labels:
    name: selenium-event-bus
    component: "selenium-grid-4"
spec:
  selector:
    app: selenium-event-bus
  ports:
    - name: port1
      protocol: TCP
      port: 4442
      targetPort: 4442
    - name: port2
      protocol: TCP
      port: 4443
      targetPort: 4443
    - name: port3
      protocol: TCP
      port: 5557
      targetPort: 5557

---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: selenium-event-bus-deployment
  labels:
    app: selenium-event-bus
    name: selenium-event-bus
    component: "selenium-grid-4"
spec:
  replicas: 1
  selector:
    matchLabels:
      app: selenium-event-bus
  template:
    metadata:
      labels:
        app: selenium-event-bus
        name: selenium-event-bus
        component: "selenium-grid-4"
    spec:
      containers:
        - name: selenium-event-bus
          image: selenium/event-bus:4.0.0-beta-4-prerelease-20210517
          ports:
            - containerPort: 4442
              protocol: TCP
            - containerPort: 4443
              protocol: TCP
            - containerPort: 5557
              protocol: TCP
          livenessProbe:
            httpGet:
              path: /status
              port: 5557
            initialDelaySeconds: 5
            periodSeconds: 5
---
#***************************************
#SELENIUM SESSIONS MAP
#***************************************
apiVersion: v1
kind: Service
metadata:
  name: selenium-sessions
  labels:
    name: selenium-sessions
    component: "selenium-grid-4"
spec:
  selector:
    app: selenium-sessions
  ports:
    - name: port1
      protocol: TCP
      port: 5556
      targetPort: 5556

---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: selenium-sessions-deployment
  labels:
    app: selenium-sessions
    name: selenium-sessions
    component: "selenium-grid-4"
spec:
  replicas: 1
  selector:
    matchLabels:
      app: selenium-sessions
  template:
    metadata:
      labels:
        app: selenium-sessions
        name: selenium-sessions
        component: "selenium-grid-4"
    spec:
      containers:
        - name: selenium-sessions
          image: selenium/sessions:4.0.0-beta-4-prerelease-20210517
          env:
            - name: SE_EVENT_BUS_HOST
              value: "selenium-event-bus"
            - name: SE_EVENT_BUS_PUBLISH_PORT
              value: "4442"
            - name: SE_EVENT_BUS_SUBSCRIBE_PORT
              value: "4443"
          ports:
            - containerPort: 5556
              protocol: TCP
          livenessProbe:
            httpGet:
              path: /status
              port: 5556
            initialDelaySeconds: 5
            periodSeconds: 5

---
#***************************************
#SELENIUM SESSION QUEUE
#***************************************
apiVersion: v1
kind: Service
metadata:
  name: selenium-session-queue
  labels:
    name: selenium-session-queue
    component: "selenium-grid-4"
spec:
  selector:
    app: selenium-session-queue
  ports:
    - name: port1
      protocol: TCP
      port: 5559
      targetPort: 5559
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: selenium-session-queue-deployment
  labels:
    app: selenium-session-queue
    name: selenium-session-queue
    component: "selenium-grid-4"
spec:
  replicas: 1
  selector:
    matchLabels:
      app: selenium-session-queue
  template:
    metadata:
      labels:
        app: selenium-session-queue
        name: selenium-session-queue
        component: "selenium-grid-4"
    spec:
      containers:
        - name: selenium-session-queue
          image: selenium/session-queue:4.0.0-beta-4-prerelease-20210517
          env:
            - name: SE_EVENT_BUS_HOST
              value: "selenium-event-bus"
            - name: SE_EVENT_BUS_PUBLISH_PORT
              value: "4442"
            - name: SE_EVENT_BUS_SUBSCRIBE_PORT
              value: "4443"
          ports:
            - containerPort: 5556
              protocol: TCP
          livenessProbe:
            httpGet:
              path: /status
              port: 5559
            initialDelaySeconds: 5
            periodSeconds: 5
---
#***************************************
#SELENIUM DISTRIBUTOR
#***************************************
apiVersion: v1
kind: Service
metadata:
  name: selenium-distributor
  labels:
    name: selenium-distributor
    component: "selenium-grid-4"
spec:
  selector:
    app: selenium-distributor
  ports:
    - name: port1
      protocol: TCP
      port: 5553
      targetPort: 5553

---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: selenium-distributor-deployment
  labels:
    app: selenium-distributor
    name: selenium-distributor
    component: "selenium-grid-4"
spec:
  replicas: 1
  selector:
    matchLabels:
      app: selenium-distributor
  template:
    metadata:
      labels:
        app: selenium-distributor
        name: selenium-distributor
        component: "selenium-grid-4"
    spec:
      containers:
        - name: selenium-distributor
          image: selenium/distributor:4.0.0-beta-4-prerelease-20210517
          env:
            - name: SE_EVENT_BUS_HOST
              value: "selenium-event-bus"
            - name: SE_EVENT_BUS_PUBLISH_PORT
              value: "4442"
            - name: SE_EVENT_BUS_SUBSCRIBE_PORT
              value: "4443"
            - name: SE_SESSIONS_MAP_HOST
              value: "selenium-sessions"
            - name: SE_SESSIONS_MAP_PORT
              value: "5556"
            - name: SE_SESSION_QUEUE_HOST
              value: "selenium-session-queue"
            - name: SE_SESSION_QUEUE_PORT
              value: "5559"
          ports:
            - containerPort: 5553
              protocol: TCP
          livenessProbe:
            httpGet:
              path: /status
              port: 5553
            initialDelaySeconds: 5
            periodSeconds: 5
---
#***************************************
#SELENIUM ROUTER
#***************************************
apiVersion: v1
kind: Service
metadata:
  name: selenium-router
  labels:
    name: selenium-router
    component: "selenium-grid-4"
spec:
  selector:
    app: selenium-router
  ports:
    - name: web
      protocol: TCP
      port: 4444
      targetPort: 4444

---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: selenium-router-deployment
  labels:
    app: selenium-router
    name: selenium-router
    component: "selenium-grid-4"
spec:
  replicas: 1
  selector:
    matchLabels:
      app: selenium-router
  template:
    metadata:
      labels:
        app: selenium-router
        name: selenium-router
        component: "selenium-grid-4"
    spec:
      containers:
        - name: selenium-router
          image: selenium/router:4.0.0-beta-4-prerelease-20210517
          env:
            - name: SE_DISTRIBUTOR_HOST
              value: "selenium-distributor"
            - name: SE_DISTRIBUTOR_PORT
              value: "5553"
            - name: SE_SESSIONS_MAP_HOST
              value: "selenium-sessions"
            - name: SE_SESSIONS_MAP_PORT
              value: "5556"
            - name: SE_SESSION_QUEUE_HOST
              value: "selenium-session-queue"
            - name: SE_SESSION_QUEUE_PORT
              value: "5559"
          ports:
            - containerPort: 4444
              protocol: TCP
          livenessProbe:
            httpGet:
              path: /status
              port: 4444
            initialDelaySeconds: 5
            periodSeconds: 5
---
#***************************************
#SELENIUM CHROME NODE
#***************************************
apiVersion: v1
kind: Service
metadata:
  name: selenium-chrome-node
  labels:
    name: selenium-chrome-node
    component: "selenium-grid-4"
spec:
  selector:
    app: selenium-chrome-node
  ports:
    - name: port1
      protocol: TCP
      port: 6900
      targetPort: 5900

---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: selenium-chrome-node-deployment
  labels:
    app: selenium-chrome-node
    name: selenium-chrome-node
    component: "selenium-grid-4"
spec:
  replicas: 28
  selector:
    matchLabels:
      app: selenium-chrome-node
  template:
    metadata:
      labels:
        app: selenium-chrome-node
        name: selenium-chrome-node
        component: "selenium-grid-4"
    spec:
      containers:
        - name: selenium-chrome-node
          image: selenium/node-chrome:90.0-chromedriver-90.0-grid-4.0.0-beta-4-prerelease-20210517
          env:
            - name: SE_EVENT_BUS_HOST
              value: "selenium-event-bus"
            - name: SE_EVENT_BUS_PUBLISH_PORT
              value: "4442"
            - name: SE_EVENT_BUS_SUBSCRIBE_PORT
              value: "4443"
          ports:
            - containerPort: 5553
              protocol: TCP
          livenessProbe:
            httpGet:
              path: /status
              port: 5555
            initialDelaySeconds: 5
            periodSeconds: 5
          volumeMounts:
            - name: dshm
              mountPath: /dev/shm
          resources:
            requests:
              memory: "2Gi"
              cpu: "1"
            limits:
              memory: "2Gi"
              cpu: "1"
      volumes:
        - name: dshm
          emptyDir: { "medium": "Memory" }
```
````

Please provide a test script to reproduce the issue you are reporting, if the 
setup is more complex, GitHub repo links are also OK.

Issues without a reproduction script are likely to stall and eventually be closed.

## Environment

Docker-Selenium image version: selenium/sessions:4.0.0-beta-4-prerelease-20210517
-->
Docker version: v1.21.1

**Comment by yasha-dev1** [comment]
Seems the problem is from the router. when I restart it, everything starts good. 
got a workaround for now with tweaking the livenessprobe of K8s: 
```
          livenessProbe:
            exec:
              command:
                - /bin/sh
                - -c
                - |
                  curl -X POST -H "Content-Type: application/json" --data '{"query": "{ grid { } }"}' --max-time 5 http://localhost:4444/graphql
            initialDelaySeconds: 5

```

**Comment by diemol** [comment]
It has been a while since this was reported and our most recent [prerelease](https://github.com/SeleniumHQ/docker-selenium/releases/tag/4.0.0-rc-1-prerelease-20210804), could you please double check this is still happening, a few memory improvements were done in the previous weeks.

Also, if it still happens, could it be possible to see what the Router resources are? When it stucks, how much memory and cpu is using? Would be nice to see a thread dump.

**Comment by yasha-dev1** [comment]
I can confirm the issue has been solved with the new prerelease (4.0.0-rc-1-prerelease-20210804). 
Closing the issue.
